src/lib.rs

```diff
@@ -107,12 +111,12 @@ pub fn encode(hash: Hash, input: &[u8]) -> Result<Multihash, EncodeError> {
 
         let total_len = code.len() + size.len() + input.len();
 
-        let mut output = BytesMut::with_capacity(total_len);
-        output.put_slice(code);
-        output.put_slice(size);
-        output.put_slice(input);
+        let mut output = Vec::with_capacity(total_len);
+        output.extend_from_slice(code);
+        output.extend_from_slice(size);
+        output.extend_from_slice(input);
         Ok(Multihash {
-            bytes: output.freeze(),
+            storage: Storage::from_slice(&output),
```

Comment:
Instead of always copying from a temporary `Vec`, how about `Storage` gets a `Storage::from_slices` method and decides internally if the accumulated length fits into its inline storage in which case we avoid the heap allocation altogether, e.g.

```rust
    pub(crate) fn from_slices(slices: &[&[u8]]) -> Self {
        let n = slices.iter().fold(0usize, |a, s| a.saturating_add(s.len()));
        if n <= MAX_INLINE {
            let s = slices.iter().fold(([0; MAX_INLINE], 0), |(mut array, i), s| {
                array[i .. i + s.len()].copy_from_slice(s);
                (array, i + s.len())
            });
            Storage::Inline(n as u8, s.0)
        } else {
            let mut v = Vec::with_capacity(n);
            for s in slices {
                v.extend_from_slice(s)
            }
            Storage::Heap(v.into())
        }
    }
```

Reply:
Yes, dealing with this temporary Vec has been on my mind, but I thought it would be best to keep it for the next PR, since it is a separate concern.

Reply:
This is only for the identity hash. Will add this and add a test.